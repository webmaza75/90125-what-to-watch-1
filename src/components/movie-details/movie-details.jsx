import React, {Fragment, PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import {
  getMoreFilmsByGenre
} from '../../utils.js';
import {
  itemShape
} from '../../models.js';
import {
  getMovieById,
  getFilms,
  getComments,
  getPlayState
} from '../../reducers/films/selectors.js';
import Header from '../header/header.jsx';
import Footer from '../footer/footer.jsx';
import FilmList from '../film-list/film-list.jsx';
import withActiveItem from '../../hocs/with-active-item/with-active-item.js';
import MovieTabs from '../movie-tabs/movie-tabs.jsx';
import {Operation} from '../../reducers/films/films.js';
import {Tabs} from '../../consts.js';
import MovieCardButtons from '../movie-card-buttons/movie-card-buttons.jsx';
import Player from '../player/player.jsx';

const FilmListWrapped = withActiveItem(FilmList);
const MovieTabsWrapped = withActiveItem(MovieTabs);

class MovieDetails extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {onLoadComments, match} = this.props;
    onLoadComments(match.params.id);
  }

  componentDidUpdate(prevProps) {
    const {onLoadComments, match} = this.props;
    if (prevProps.match.params.id !== match.params.id) {
      onLoadComments(match.params.id);
    }
  }

  render() {
    const {
      movie,
      films,
      comments,
      location,
      showPlayer
    } = this.props;

    if (!movie || !films) {
      return null;
    }

    const {
      id,
      backgroundImage,
      genre,
      title,
      posterImage,
      released,
      isFavorite
    } = movie;
    const moreFilms = getMoreFilmsByGenre(films, movie);

    if (showPlayer) {
      return <Player movie={movie} />;
    }

    return <Fragment>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={backgroundImage} alt={title} />
          </div>

          <h1 className="visually-hidden">WTW</h1>
          <Header />

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{title}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{released}</span>
              </p>

              <MovieCardButtons
                id={id}
                isFavorite={isFavorite}
                showAddReviewLink={true}
              />

            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={posterImage} alt={`${title} poster`} width="218" height="327" />
            </div>

            <MovieTabsWrapped
              activeItem ={Tabs.OVERVIEW}
              href={`/film/${id}`}
              key={id}
              location={location}
              movie={movie}
              reviews={comments}
            />

          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <FilmListWrapped
            films={moreFilms}
          />
        </section>
        <Footer />
      </div>
    </Fragment>;
  }
}

MovieDetails.propTypes = {
  movie: itemShape,
  films: PropTypes.arrayOf(itemShape),
  location: PropTypes.object,
  onLoadComments: PropTypes.func,
  comments: PropTypes.arrayOf(PropTypes.object),
  match: PropTypes.object,
  history: PropTypes.object,
  showPlayer: PropTypes.bool
};

const mapStateToProps = (state, {match}) => ({
  movie: getMovieById(state, match.params.id),
  films: getFilms(state),
  comments: getComments(state),
  showPlayer: getPlayState(state)
});

const mapDispatchToProps = {
  onLoadComments: Operation.loadComments
};

export {MovieDetails};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MovieDetails);
