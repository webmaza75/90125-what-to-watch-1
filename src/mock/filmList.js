const filmList = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`,
  `Aviator`
];

export default filmList;
