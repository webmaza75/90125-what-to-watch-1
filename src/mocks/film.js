const film = {
  genre: [`Drama`, `Thriller`],
  title: `Aviator`,
  desc: ``,
  picture: `aviator.jpg`,
  year: 2016,
  src: ``
};

export default film;
