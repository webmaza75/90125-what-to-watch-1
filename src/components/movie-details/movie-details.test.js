import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import {MovieDetails} from './movie-details.jsx';
import film from '../../mocks/film.js';
import filmList from '../../mocks/films.js';

it(`MovieDetails correctly renders`, () => {
  const renderer = new ShallowRenderer();
  const result = renderer
    .render(<MovieDetails
      movie={film}
      films={filmList}
      comments={[]}
      isAuthorized={false}
      onLoadComments={jest.fn()}
      onToggleFavorite={jest.fn()}
    />);

  expect(result).toMatchSnapshot();
});
