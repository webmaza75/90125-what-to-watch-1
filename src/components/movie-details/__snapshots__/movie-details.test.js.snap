// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails correctly renders 1`] = `
<React.Fragment>
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="War of the Worlds"
          src="https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <Memo(ConnectFunction) />
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            War of the Worlds
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Adventure
            </span>
            <span
              className="movie-card__year"
            >
              2005
            </span>
          </p>
          <Buttons
            id={1}
            isAuthorized={false}
            isFavorite={false}
            onMyListClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="War of the Worlds poster"
            height="327"
            src="https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
            width="218"
          />
        </div>
        <WithActiveItem
          activeItem="Overview"
          href="/film/1"
          movie={
            Object {
              "backgroundColor": "#9B7E61",
              "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg",
              "description": "As Earth is invaded by alien tripod fighting machines, one family fights for survival.",
              "director": "Steven Spielberg",
              "genre": "Adventure",
              "id": 1,
              "isFavorite": false,
              "picture": "https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg",
              "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg",
              "rating": 6.5,
              "ratingLevel": "Good",
              "released": 2005,
              "runTime": 116,
              "scoresCount": 386834,
              "src": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
              "starring": Array [
                "Tom Cruise",
                "Dakota Fanning",
                "Tim Robbins",
              ],
              "title": "War of the Worlds",
              "videoLink": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
            }
          }
          reviews={Array []}
        />
      </div>
    </div>
  </section>
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <WithActiveItem
          films={
            Array [
              Object {
                "backgroundColor": "#D8E3E5",
                "backgroundImage": "https://es31-server.appspot.com/wtw/static/film/background/Moonrise_Kingdom.jpg",
                "description": "A pair of young lovers flee their New England town, which causes a local search party to fan out to find them.",
                "director": "Wes Anderson",
                "genre": "Adventure",
                "id": 3,
                "isFavorite": false,
                "picture": "https://es31-server.appspot.com/wtw/static/film/preview/moonrise-kingdom.jpg",
                "posterImage": "https://es31-server.appspot.com/wtw/static/film/poster/Moonrise_Kingdom.jpg",
                "rating": 7.8,
                "ratingLevel": "Good",
                "released": 2012,
                "runTime": 94,
                "scoresCount": 291183,
                "src": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "starring": Array [
                  "Jared Gilman",
                  "Kara Hayward",
                  "Bruce Willis",
                ],
                "title": "Moonrise Kingdom",
                "videoLink": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
              },
            ]
          }
        />
      </div>
    </section>
    <Footer />
  </div>
</React.Fragment>
`;
